sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,315 @@
+# -*- coding: utf-8 -*-
+"""K-medoids clustering"""
+
+# Authors: Timo Erkkilä <timo.erkkila@gmail.com>
+#          Antti Lehmussola <antti.lehmussola@gmail.com>
+#          Kornel Kiełczewski <kornel.k@plusnet.pl>
+# License: BSD 3 clause
+
+import warnings
+
+import numpy as np
+
+from ..base import BaseEstimator, ClusterMixin, TransformerMixin
+from ..metrics.pairwise import PAIRWISE_DISTANCE_FUNCTIONS
+from ..utils import check_array, check_random_state
+from ..utils.validation import check_is_fitted
+
+
+class KMedoids(BaseEstimator, ClusterMixin, TransformerMixin):
+    """
+    k-medoids class.
+
+    Parameters
+    ----------
+    n_clusters : int, optional, default: 8
+        How many medoids. Must be positive.
+
+    distance_metric : string, optional, default: 'euclidean'
+        What distance metric to use.
+
+    clustering : {'pam'}, optional, default: 'pam'
+        What clustering mode to use.
+
+    init : {'random', 'heuristic'}, optional, default: 'heuristic'
+        Specify medoid initialization.
+
+    max_iter : int, optional, default : 300
+        Specify the maximum number of iterations when fitting.
+
+    random_state : int, optional, default: None
+        Specify random state for the random number generator.
+    """
+
+    # Supported clustering methods
+    CLUSTERING_METHODS = ['pam']
+
+    # Supported initialization methods
+    INIT_METHODS = ['random', 'heuristic']
+
+    def __init__(self, n_clusters=8, distance_metric='euclidean',
+                 clustering_method='pam', init='heuristic',
+                 max_iter=300, random_state=None):
+
+        self.n_clusters = n_clusters
+
+        self.distance_metric = distance_metric
+
+        self.init = init
+
+        self.max_iter = max_iter
+
+        self.clustering_method = clustering_method
+
+        self.random_state = random_state
+
+    def _check_init_args(self):
+
+        # Check n_clusters
+        if self.n_clusters is None or self.n_clusters <= 0 or \
+                not isinstance(self.n_clusters, int):
+            raise ValueError("n_clusters has to be nonnegative integer")
+
+        # Check distance_metric
+        if callable(self.distance_metric):
```

Comment:
I think you can avoid this validation if you use the `metrics.pairwise_distances` function...